spi: add SPI driver for STM32 family

superna9999 opened the pull request:
Add a SPI master and slave driver for the STM32 SoCs.

Originally submitted on Gerrit: https://gerrit.zephyrproject.org/r/#/c/12286/6 for STm32Lx as native driver.

Changes since gerrit revision 6:
 - Converted uint*_t to u*_t

Then @ldts and @erwango helped move to ST HAL to have a LL based driver for all STM32 SoCs.

superna9999 committed:
spi: add SPI driver for STM32 family

Add a SPI master and slave driver for the L4, F4 and F3 STM SoCs families.

Change-Id: I1faf5c97f992c91eba852fd126e7d3b83158993d
Origin: Original
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>
Tested-by: Lee Jones <lee.jones@linaro.org>

superna9999 committed:
pinmux: stm32: nucleo_l476rg: Fix SPI Pinmux

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
boards: nucleo_l476rg: Document default SPI pinmux

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

nashif commented:
rebase and resubmit please, the ARM branch was out of date

superna9999 commented:
Hi, @galak, is there something missing for this driver ?

MaureenHelm commented on drivers/spi/spi_stm32lx.c:
```diff
@@ -0,0 +1,398 @@
+	}
+
+	for (br = 0 ; br < BR_SCALER_COUNT ; ++br) {
+		u32_t clk = clock / baud_rate_scaler[br];
```
Since baud_rate_scaler is always a power of two, it would be more efficient to do a shift instead of a divide

superna9999 replied:
ok

MaureenHelm commented on drivers/spi/spi_stm32lx.c:
```diff
@@ -0,0 +1,398 @@
+}
+#endif
+
+#define BR_SCALER_COUNT		8
```
It would be safer to define this after baud_rate_scaler and use the ARRAY_SIZE() macro

superna9999 replied:
ok

superna9999 replied:
I removed the table since the br matches the >> value.

MaureenHelm commented on drivers/spi/spi_stm32lx.c:
```diff
@@ -0,0 +1,398 @@
+
+	return 0;
+}
+
```
Should there be a SPI_0 instance here?

superna9999 replied:
No, @erwango insists on having a 1:1 relationship with the numbering scheme used in the ST documentation
It's already the case for UART and I2C drivers.

erwango replied:
Confirmed

MaureenHelm commented on drivers/spi/spi_stm32lx.c:
```diff
@@ -0,0 +1,398 @@
+
+DEVICE_AND_API_INIT(spi_stm32lx_1, CONFIG_SPI_1_NAME, &spi_stm32lx_init,
+		    &spi_stm32lx_dev_data_1, &spi_stm32lx_cfg_1,
+		    PRE_KERNEL_2, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
```
Does this need to be initialized in PRE_KERNEL_2? Can we postpone to POST_KERNEL?

superna9999 replied:
We can, it's a copy paste issue.

superna9999 replied:
done

tbursztyka commented on drivers/spi/Kconfig:
```diff
@@ -188,6 +203,44 @@ config SPI_2_CS_GPIO_PIN
+	int "Port 3 default baud rate"
+	depends on SPI_3
+	default 500000
+
```
remove these 2 options above as you don't seem to use it. This was introduced by qmsi spi shim driver, for no good reason actually (and will become obsolete as soon as new spi API gets in)

tbursztyka commented on drivers/spi/spi_stm32lx.c:
```diff
@@ -0,0 +1,398 @@
+	int br;
+
+	clock_control_get_rate(data->clock,
+			(clock_control_subsys_t *)&cfg->pclken, &clock);
```
parameters alignement:
foo(param1, ...
      paramn, ... );

Seems like you could break a line before &clock to get it

This rules applies everywhere on function declaration and call, please apply it where relevant, I see many places where it could be applied.

superna9999 replied:
ok, done

tbursztyka commented on drivers/spi/spi_stm32lx.c:
```diff
@@ -0,0 +1,398 @@
+	/* NOP */
+
+	return 0;
+}
```
you can remove that function altogether and not fill in a value in api_funcs, if NULL the public api will return 0

superna9999 replied:
done

ldts commented:
@superna9999 , have you tried hardwiring MISO/MOSI lines - leaving NSS open -  and executing a loopback test? I am testing on nucleo and carbon that way with an old SPI driver we did for Zephyr last year and noticed a change seems to be needed in pinmux . Not sure if it is a problem in our old driver so I thought I'd check with you the behavior of this new driver.

superna9999 commented:
@ldts Let me check back, but we use this driver in a near production ready product, so I assume it should work

superna9999 commented:
@ldts I tested on a Nucle_L432KC board and it worked fine

I fixed a small issue when CS is not exposed on a pin, but MISO input was good when connected to MOSI.

ldts commented:
@superna9999  thanks! I'll pull again and retest.
just so you know, I am using your driver on nucleo_401re and works as expected.

erwango committed:
stm32cube: Fix warning when SPI LL API is compiled

Current implementation of LL_SPI_TransmitData16 on F3/F7/L4 family
generates following warning:
"warning: dereferencing type-punned pointer will break strict-aliasing
rules [-Wstrict-aliasing]"
Besides being forbidden by rule, this cast is not needed, as register is
16 bits wide. Modification has been tested on L4 SoC.
stm32yyxx_ll_spi.h being included in soc.h file, warning is generated
at each compiled object, this commit allows a clean build.
This issue is referenced in ST and tracked under
reference  13359. Code will be updated on upcoming stm32cube updates.

Change-Id: I3ca54a81d849d4852eca86b52b6825b60e18b752
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

erwango committed:
pinmux: stm32: nucleo_f334r8: add support for SPI

Following implementation of LL based SPI driver,
add SPI support on nucleo_f334r8 board of STM32F3 series.

Change-Id: Ifbe39b1f2cecdd7db23be9c6943a914a155ebd77
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

ldts committed:
pinmux: stm32f4: Add SPI2 pins on PB12, PB13, PB14 & PB15

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

ldts committed:
pinmux: stm32: nucleo_f401re: Add support for SPI

Add SPI pin for the nucleo_f401re pinmux.

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

ldts committed:
stm32cube: build stm32xxx_ll_spi if CONFIG_SPI

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

ldts committed:
pinmux: stm32f4: Add SPI1 pins on PA4, PA5, PA6 & PA7

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
pinmux: stm32: nucleo_l432kc: Add SPI pins

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

ldts commented:
@superna9999 just checking that you are still ok with upstreaming your driver with the LL API instead of the original native version? I am only asking in case I need to start tracking :) TIA!

superna9999 commented:
@ldts yes, I still need to do 2/3 changes and I'll switch this PR to it

Commit:
pinmux: pinmux_dev_k64 driver and related references are removed.

Functionality of a pinmux driver is now a part of a regular
driver.

Signed-off-by: Shiksha Patel <shiksha.patel@nxp.com>

jamike committed:
flash: stm32: fix for l4 writing wrong data

L4 have 64 bits write access.
The cast to 64 bits data address in write_dword
requires 3 right shifts on i (byte index) else the
data taken are wrong for i different from 0

Signed-off-by: Michel Jaouen <michel.jaouen@st.com>

galak committed:
arm: Add build time consistency check for irq priority defines

We need to make sure that __NVIC_PRIO_BITS & CONFIG_NUM_IRQ_PRIO_BITS
are set to the same value.  Add a simple build time check to ensure
this is the case.  This is to catch future cases of issues like
ZEP-2243.  This is a stop gap til we resolve ZEP-2262, which covers use
of both __NVIC_PRIO_BITS & CONFIG_NUM_IRQ_PRIO_BITS.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
boards: 96b_nitrogen: Add support for flash/debug with pyOCD

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: bossa-flash.sh: fix variable usage

We define a variable to pickup a default for the bossa binary, however
we weren't using it.  Lets do so now.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

MaureenHelm committed:
arm: nxp: mpu: Fix region descriptor 0 attributes

Clearing fields in the region descriptor attributes doesn't always have
the expected effect of revoking permissions. In the case of bus master
supervisor mode fields (MxSM), setting to zero actually enables read,
write, and execute access.

When we reworked handling of region descriptor 0, we inadvertently
enabled execution from RAM by clearing the MxSM fields and enabling the
descriptor. This caused samples/mpu_test run to throw a usage fault
instead of an MPU-triggered bus fault.

Fix this by setting all the MxSM fields to 2'b11, which gives supervisor
mode the same access as user mode.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

jamike committed:
flash: stm32: distinguish read/write for flash range valid

L4 write access requires 64 bits alignment
while L4 read access does not require any alignment.
To support specific check according to read/write,erase
a parameter is added to stm32_valid_range.

Signed-off-by: Michel Jaouen <michel.jaouen@st.com>

mnkp committed:
arch: same70: Fix ERASE pin configuration

Correct code that allows to disable ERASE pin functionality
during boot.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
boards: sam_e70_xplained: allow flashing via JTAG header

Allow to use an external debug adapter such as J-Link or ULINK
connected to a 20-pin JTAG header to flash the image. SWD is
the actual protocol used by the debug interface.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

Commit:
arm: Modify linker script to accomodate need for flash footer.

The porting of the TI CC2650 SoC introduces the need to
write a specific configuration area (CCFG) at the end of the
flash. It is read by the bootloader ROM of the SoC.

For now, this is a quick hack and not a generic solution;
similar needs may arise with other hardware.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
drivers: serial: uart_stellaris: Remove UART_IRQ_FLAGS

We always have UART_IRQ_FLAGS set to 0, so just call IRQ_CONNECT with a
0 argument for the flags, and remove the UART_IRQ_FLAGS.  This is
towards support for using the driver on the TI CC2650.  (we add a
comment about that as well).

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: pyocd.sh: Add support for passing board_id to pyocd commands

If we have more than one board of a given type we need to be able to
specify the board_id to select which specific board that the pyocd
command should target.  Introduce an environment variable PYOCD_BOARD_ID
to we can set that will get passed to the pyocd command that needs it.

Here's an example:

$ make -C samples/hello_world/ BOARD=frdm_k64f flash PYOCD_BOARD_ID=1234

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
board: frdm_k64f: allow overriding default debug/flash scripts

As there are multiple ways to flash or debug (pyOCD or openOCD) allow
the user to override the default.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

Commit:
uart: Use DTS labels for Stellaris driver.

Update driver to use DTS-generated #defines for port names,
and not obsolete Kconfig variables.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
arm: Add support for TI's CC2650 SoC.

Add support in arch/arm/soc/ti_simplelink, along with support
for CC32xx SoC.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
sensortag: Add TI's SensorTag board.

Add support for TI's SensorTag board, which uses a CC2650 SoC.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
cc2650: Add GPIO driver.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
cc2650: Add pinmux driver.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
samples: gpio: Add support for SensorTag board.

Add extra #defines in samples/drivers/gpio to test
TI SensorTag board.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>
